lib/active_model/serializer/attributes.rb

```diff
@@ -1,55 +1,23 @@
 module ActiveModel
   class Serializer
     module Attributes
-      # @api private
-      class Attribute
-        delegate :call, to: :reader
-
-        attr_reader :name, :reader
-
-        def initialize(name)
-          @name = name
-          @reader = :no_reader
-        end
-
-        def self.build(name, block)
-          if block
-            AttributeBlock.new(name, block)
-          else
-            AttributeReader.new(name)
-          end
-        end
-      end
-      # @api private
-      class AttributeReader < Attribute
-        def initialize(name)
-          super(name)
-          @reader = ->(instance) { instance.read_attribute_for_serialization(name) }
-        end
-      end
-      # @api private
-      class AttributeBlock < Attribute
-        def initialize(name, block)
-          super(name)
-          @reader = ->(instance) { instance.instance_eval(&block) }
-        end
-      end
-
       extend ActiveSupport::Concern
 
       included do
         with_options instance_writer: false, instance_reader: false do |serializer|
-          serializer.class_attribute :_attribute_mappings # @api private : maps attribute key names to names to names of implementing methods, @see #attribute
-          self._attribute_mappings ||= {}
+          serializer.class_attribute :_attributes_data # @api private
+          self._attributes_data ||= {}
```

Comment:
So you're using a hash to ensure unique attributes via its lookup keys, but otherwise only using the values?

Interesting. I believe that's what Set does internally, or functionally similar.

What do you see as the downside of just declaring the key in the serializer's data structure and let the attribute just be the value (ie previous vs current)?

B mobile phone

> On Dec 28, 2015, at 1:49 PM, Lucas Hosseini <notifications@github.com> wrote:
> 
> In lib/active_model/serializer/attributes.rb:
> 
> >        extend ActiveSupport::Concern
> >  
> >        included do
> >          with_options instance_writer: false, instance_reader: false do |serializer|
> > -          serializer.class_attribute :_attribute_mappings # @api private : maps attribute key names to names to names of implementing methods, @see #attribute
> > -          self._attribute_mappings ||= {}
> > +          serializer.class_attribute :_attributes_data # @api private
> > +          self._attributes_data ||= {}
> Oh I see. Thing is, _attributes_data could really be just an array, but access indexed by attribute name makes it easier to avoid/catch duplicate attribute definitions (although it shouldn't happen).
> 
> —
> Reply to this email directly or view it on GitHub.
>